jax/version.py

```diff
@@ -14,18 +14,103 @@
 
 # This file is included as part of both jax and jaxlib. It is also
 # eval()-ed by setup.py, so it should not have any dependencies.
+from __future__ import annotations
 
-import os
 import datetime
+import os
+import pathlib
+import subprocess
 
 _version = "0.4.16"
+# The following line is overwritten by build scripts in distributions &
+# releases. Do not modify this manually, or jax/jaxlib build will fail.
+_release_version: str | None = None
+
+
+def _get_version_string() -> str:
+  # The build/source distribution for jax & jaxlib overwrites _release_version.
+  # In this case we return it directly.
+  if _release_version is not None:
+    return _release_version
+  try:
+    return _version_from_git_tree(_version)
+  except:
+    return _version_from_todays_date(_version)
+
+
+def _version_from_todays_date(base_version: str) -> str:
+  datestring = datetime.date.today().strftime("%Y%m%d")
+  return f"{base_version}.dev{datestring}"
+
+
+def _version_from_git_tree(base_version: str) -> str:
+  stdout = subprocess.check_output(["git", "show", "-s", "--format=%at", "HEAD"])
+  timestamp = int(stdout.decode().strip())
+  datestring = datetime.date.fromtimestamp(timestamp).strftime("%Y%m%d")
+  assert datestring.isnumeric()
+
+  stdout = subprocess.check_output(["git", "describe", "--long", "--always"])
+  commit_hash = stdout.decode().strip().rsplit('-', 1)[-1]
+  assert commit_hash.isalnum()
+
+  return f"{base_version}.dev{datestring}+{commit_hash}"
+
+
+def get_version_for_build() -> str:
```

Comment:
Is this name is exported from the public JAX namespace? We'd want to avoid that, ideally.